front/components/data_source_view/DataSourceBuilderSelector.tsx

```diff
@@ -0,0 +1,242 @@
+ */
+const useNavigationHistory = () => {
+  const [navigationHistory, setHistory] = useState<
+    NavigationHistoryEntryType[]
+  >([{ type: "root" }]);
+
+  const setSpace = useCallback((space: SpaceType) => {
+    setHistory((prev) => {
+      return [...prev.slice(0, 1), { type: "space", space }];
+    });
+  }, []);
+
+  const setCategory = useCallback(
+    (category: DataSourceViewCategoryWithoutApps) => {
+      setHistory((prev) => {
+        return [...prev.slice(0, 2), { type: "category", category }];
+      });
+    },
+    []
+  );
+
+  const addNode = useCallback((node: DataSourceViewContentNode) => {
+    setHistory((prev) => {
+      return [...prev, { type: "node", node }];
+    });
+  }, []);
+
+  const navigateTo = useCallback((index: number) => {
+    setHistory((prev) => prev.slice(0, index + 1));
+  }, []);
+
+  return {
+    navigationHistory,
+    setSpace,
+    setCategory,
+    addNode,
+    navigateTo,
+  };
+};
+
+interface DataSourceBuilderSelectorProps {
+  allowedSpaces?: SpaceType[];
+  owner: LightWorkspaceType;
+  dataSourceViews: DataSourceViewType[];
+  selectionConfigurations: DataSourceViewSelectionConfigurations;
+  setSelectionConfigurations: Dispatch<
+    SetStateAction<DataSourceViewSelectionConfigurations>
+  >;
+  viewType: ContentNodesViewType;
+}
+
+export const DataSourceBuilderSelector = ({
+  allowedSpaces,
+  dataSourceViews,
+  owner,
+  viewType,
+}: DataSourceBuilderSelectorProps) => {
+  const { spaces, isSpacesLoading } = useSpaces({ workspaceId: owner.sId });
+
+  const { navigationHistory, setSpace, setCategory, addNode, navigateTo } =
+    useNavigationHistory();
+  const currentNavigationEntry =
+    navigationHistory[navigationHistory.length - 1];
+
+  const selectedSpaceId =
+    findSpaceFromNavigationHistory(navigationHistory)?.sId;
+  const selectedCategory = findCategoryFromNavigationHistory(navigationHistory);
+  const traversedNode = getLatestNodeFromNavigationHistory(navigationHistory);
+
+  const [searchQuery, setSearchQuery] = useState("");
+
+  // Fetch category data when a category is selected
+  const {
+    spaceDataSourceViews: categoryDataSourceViews,
+    isSpaceDataSourceViewsLoading: isCategoryLoading,
+  } = useSpaceDataSourceViewsWithDetails({
+    workspaceId: owner.sId,
+    spaceId: selectedSpaceId || "",
+    category: selectedCategory || "managed",
+    disabled: !selectedSpaceId || !selectedCategory,
+  });
+
+  const breadcrumbItems: BreadcrumbItem[] = useMemo(
+    () =>
+      navigationHistory.map((entry, index) => ({
+        ...getBreadcrumbConfig(entry),
+        href: undefined,
+        onClick: () => navigateTo(index),
+      })),
+    [navigationHistory, navigateTo]
+  );
+
+  // Filter spaces to only those with data source views
+  const filteredSpaces = useMemo(() => {
+    const spaceIds = new Set(dataSourceViews.map((dsv) => dsv.spaceId));
+    return spaces.filter((s) => spaceIds.has(s.sId));
+  }, [spaces, dataSourceViews]);
```

Comment:
Not an expert yet, but here I don't think the useMemo is relevant, both dependencies are stables.

Reply:
They could change after a browser focus, swr will refetch spaces, dataSources as well but it's then coming from the `AgentBuilderProvider` so it would re-render the whole `DataSourceBuilderSelector` in any case.
I'd kept that mainly to avoid doing the same filtering operation each time an other state update (like `navigationHistory` or `searchQuery`)